syft/pkg/cataloger/java/archive_parser.go

```diff
@@ -230,44 +225,42 @@ func (j *archiveParser) packagesFromPomProperties(propsObj *pkg.PomProperties, p
 	matchesParentPkg = matchesParentPkg || parentPkg.Metadata.(pkg.JavaMetadata).VirtualPath == virtualPath
 
 	// the pom artifactId has the parent name or vice versa
-	if propsObj.ArtifactID != "" {
-		matchesParentPkg = matchesParentPkg || strings.Contains(parentPkg.Name, propsObj.ArtifactID) || strings.Contains(propsObj.ArtifactID, parentPkg.Name)
+	if pomProperties.ArtifactID != "" {
+		matchesParentPkg = matchesParentPkg || strings.Contains(parentPkg.Name, pomProperties.ArtifactID) || strings.Contains(pomProperties.ArtifactID, parentPkg.Name)
 	}
 
-	if matchesParentPkg {
-		// we've run across more information about our parent package, add this info to the parent package metadata
-		// the pom properties is typically a better source of information for name and version than the manifest
-		if parentPkg.Name == "" {
-			parentPkg.Name = p.Name
-		}
-		if parentPkg.Version == "" {
-			parentPkg.Version = p.Version
-		}
+	if !matchesParentPkg {
+		// only keep packages we haven't seen yet (and are not related to the parent package)
+		return []pkg.Package{p}
+	}
 
-		// keep the pom properties, but don't overwrite existing pom properties
-		parentMetadata, ok := parentPkg.Metadata.(pkg.JavaMetadata)
-		if ok && parentMetadata.PomProperties == nil {
-			parentMetadata.PomProperties = propsObj
-			parentPkg.Metadata = parentMetadata
-		}
+	// we've run across more information about our parent package, add this info to the parent package metadata
+	// the pom properties is typically a better source of information for name and version than the manifest
+	if parentPkg.Name == "" {
+		parentPkg.Name = p.Name
+	}
+	if parentPkg.Version == "" {
+		parentPkg.Version = p.Version
 	}
 
-	if !matchesParentPkg && !j.discoveredPkgs.Contains(pkgKey) {
-		// only keep packages we haven't seen yet (and are not related to the parent package)
-		pkgs = append(pkgs, p)
+	// We may have learned more about the type via data in the pom properties
+	parentPkg.Type = p.Type
+
+	// keep the pom properties, but don't overwrite existing pom properties
+	parentMetadata, ok := parentPkg.Metadata.(pkg.JavaMetadata)
+	if ok && parentMetadata.PomProperties == nil {
+		parentMetadata.PomProperties = &pomProperties
+		parentPkg.Metadata = parentMetadata
 	}
-	return pkgs
+
+	return nil
 }
 
 // discoverPkgsFromNestedArchives finds Java archives within Java archives, returning all listed Java packages found and
 // associating each discovered package to the given parent package.
 func (j *archiveParser) discoverPkgsFromNestedArchives(parentPkg *pkg.Package) ([]pkg.Package, error) {
 	var pkgs = make([]pkg.Package, 0)
 
-	if !j.detectNested {
```

Comment:
nit question: why move this from here? the original code makes it more difficult to misuse this function relative to the `archiveParser` configuration.

Reply:
Good question. The value of `j.detectNested` is always `true` in the application code. Because of that, I believe we could consider removing it in the future, to reduce the number of variables needing to be consulted in the logic.

In the test code, it looks like the value of `j.detectNested` is always `false`. I think this warrants some discussion. As it reads now, it looks like that parameter exists to facilitate testing, although it's never used with the same value in testing (`false`) as it is in application code (`true`). If it's configurable in the application code, we should consider testing all states in the test code. And, if it's not needed in the application code, we should consider whether or not it's needed in the test code. I was wondering if it was used in testing out of convenience, so we can reuse nested archives for tests that only need to expect a single package. I could probably benefit from learning more of the context here.

In the meantime, I wanted to surface this check so the conditionality was more obvious to the caller, and so that we avoid the call to `discoverPkgsFromNestedArchives` if there's not a need to invoke it. (I'll also admit, **_subjectively_**, this was reading better to me: "Do we need to get nested archives? Yes? Okay, then let's do it." vs. "Get me the nested archives. (Okay, but we might not actually.)" — again, subjective 😃 )

For this particular PR, I can move the check back if that's more palatable, and we can follow up with a discussion after the PR is merged.